### PR TITLE
feat: add exif based image rotation

### PR DESCRIPTION
**Pull request by mryel00**
This adds a basic exif based image rotation to solve #33.
Sending the exif header with the desired [rotation](https://exiftool.org/TagNames/EXIF.html) before the frame will rotate the frame in all exif supporting browsers.
This adds the new cli parameter --orientation_exif and requires an integer based on the [rotation](https://exiftool.org/TagNames/EXIF.html) desired.

In it's current form it always attaches the header, if the parameter is set. So there could be a case were the header is already there from the camera.

Signed-off-by: Patrick Gehrsitz <58853838+mryel00@users.noreply.github.com>

### CODE REVIEW

**Comment by roamingthings** [comment]
I've decoded the EXIF header generated by your tool manually and uploaded in a separate [branch](https://github.com/roamingthings/spyglass/blob/f134c9aa153ff5b250a5749a30e8f1e4a83f98e6/spyglass/server.py#L93)
Next I will try to remove everything exception the orientation.

**Comment by roamingthings** [comment]
@mryel00 can I update the PR with my branch? I would also like to move the header generation to a separate file for better testing.

**Comment by mryel00** [comment]
@roamingthings do whatever you want 😄

**Comment by Caouette1988** [comment]
For your info: I'm using this branch right now for 1 week with no problems.

**Comment by roamingthings** [comment]
@Caouette1988 I updated this branch and would appreciate if you give it a try

**Comment by roamingthings** [comment]
@mryel00 I've update the implementation. I'm not happy with the test setup I had to do. I will give it a closer look. However, the implementation of the feature should be fine.

File: spyglass/exif.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        # v-- YCbCrPositioning tag (tag number = 0x0213, type = USHORT, count = 1)
+        b'\x02\x13', b'\x00\x03', b'\x00\x00\x00\x01',
+        b'\x00\x01\x00\x00',  # center of pixel array
+        b'\x00\x00\x00\x00',  # Offset to next IFD 0
+        b'\x00\x00\x00\x48\x00\x00\x00\x01',  # XResolution value
+        b'\x00\x00\x00\x48\x00\x00\x00\x01'  # YResolution value
+    ])
+
+
+option_to_exif_orientation_map = {
```
Comment (mryel00) [inline spyglass/exif.py:38]: The correct term in python would be `dictionary` and not `map`

Comment (roamingthings) [reply spyglass/exif.py:38]: I thought about it mor from a concept point of view not a technical. It maps from one to the other.

Comment (mryel00) [reply spyglass/exif.py:38]: That makes sense ofc.
Dictionary would be still a suitable term in the conceptual point of view as you are using a dictionary to understand the meaning of a word/term.
Feel free to resolve this conversation if you want to stay with map.

**Comment by Caouette1988** [comment]
I just try to install it and it return some error.

1) I'm only able to make it work by sending this command : ./[run.py](http://run.py/) -b 0.0.0.0 -p 8080 -r 1536x864 -f 5 -st '/?action=stream' -sn '/?action=snapshot' -af continuous -l 0.0 -s normal -or r90

2) Is there an errir in spyglass.conf? It is still calling for ORIENTATION_EXIF="1" but it seem to want R90 instead of 1. I try it with no sucess.

3) spyglass.conf will need to be udated with these value :
                             '  h      - Horizontal (normal)\n'
                             '  mh     - Mirror horizontal\n'
                             '  r180   - Rotate 180\n'
                             '  mv     - Mirror vertical\n'
                             '  mhr270 - Mirror horizontal and rotate 270 CW\n'
                             '  r90    - Rotate 90 CW\n'
                             '  mhr90  - Mirror horizontal and rotate 90 CW\n'
                             '  r270   - Rotate 270 CW'

**Comment by mryel00** [comment]
@Caouette1988 thank you for spotting that error. It should be fixed now

**Comment by Caouette1988** [comment]
> @Caouette1988 thank you for spotting that error. It should be fixed now

i install it yesterday, all work good. Obico too is working.

**Comment by roamingthings** [comment]
@mryel00 I think this is ready for release. What do you think?

**Comment by mryel00** [comment]
@roamingthings I'm not sure if we need something like a fail check for the first image to check if there is already an EXIF header. I guess this might be just a simple check for the EXIF identifier in the first few bytes but I'm not completely sure about that and if we should really implement it as there shouldn't be one in nearly all scenarios.

**Comment by roamingthings** [comment]
@mryel00 This would mean that picamera2 starts to embed EXIF data. I don't think that this will be an issue.

**Comment by roamingthings** [comment]
Well I think we would have to check the first few bytes for being an EXIF header. Shouldn't be that hard.

**Review by roamingthings — APPROVED** [review]
Let's go

**Pull request merged**